### PR TITLE
CMEK support changes

### LINKED ISSUE #78: Enhancement to support CMEK Encrypted Disks
**What would you like to be added**:
1. End to end changes required to support CMEK Encrypted disks. See source requirement at: https://github.com/gardener/gardener-extension-provider-gcp/issues/564 

**Why is this needed**:
To support  https://github.com/gardener/gardener-extension-provider-gcp/issues/564  for ITAR

####  Machine Class Changes
- [x] Addition of `kmsKeyName` and in the provider spec to map inside `kmsKeyName` in GCP `CustomerEncryptionKey`
   Machine Controller Manager GCP Provider Changes. There is no need for separate specification of `kmsKeyServiceAccountName` since the current `GCPProviderSpec` already has a `ServiceAccounts []GCPServiceAccount` and introducing another would be redundant. (NOTE: Clarify why is this a slice since technically a VM can be associated with only ONE service account)
- [x] Enhancement of `GCPProviderSpec.GCPDisk` with kms key name field. 
- [x] Enhancement of `GCPDisk` such that [AttachedDisks](https://pkg.go.dev/google.golang.org/api/compute/v1#AttachedDisk) are populated using the right encryption key using key name from the machine class, whose origin source is in the shoot yaml.
- [x] Enhancement of `CreateMachine->CreateMachineUtil` to ensure that `compute.Instances` are launched with encrypted disks populating the `CustomerEncryptionKey` struct appropriately using the `kmsKeyName`  and optionally `kmsKeyServiceAccountName` taken from the `GCPProviderSpec`.
- [x] Unit tests
- [x] End to End tests

#### GCP Provider Extension
- [x] Enhance `github.com/gardener/gardener-extension-provider-gcp/pkg/apis/gcp.WorkerConfig` in backward incompatible manner to add fields for both `bootVolume` and `dataVolume`, each containing encryption key names: `kmsKeyName`
- [x] Enhance the machine deployment generation code in `generateMachineConfig`  to generate machine classes with `kmsKeyName` and optionally `kmsKeyServiceAccountName` in the provider spec.
- [x] Unit Tests
- [x] End to End Tests
- [x] Resolve addition of IAM Policy bindings to keys wrt service accounts
- [x] Resolve on whether whether `roles/cloudkms.admin` or `roles/cloudkms.cryptoKeyEncrypterDecrypter` is sufficient. FINDING:  `roles/cloudkms.cryptoKeyEncrypterDecrypter` is sufficient for encrypted disks. `roles/cloudkms.admin` is only needed for creation/deletion and general admin of keys.

#### Timeline

- [x] Gardener extension release (22nd June)
- [x] MCM-provider GCP (16 June) (Done on 8th June)

Live update (28th June) , Canary Freeze (22nd June)

Target the releases from our side -> 16th June

##### Multiple GCP Service Accounts
- [x] Test VM launch with multiple GCP service accounts.
- [x] Test diff KMS keys for diff disks giving privilege to these GCP service accounts. 
- [x] Test with old shoot YAMLs and new shoot YAMLs. Backward compat testing

### PR DESCRIPTION
**Pull request by elankath**
**What this PR does / why we need it**:
Changes required to support CMEK Encrypted disks in GCP.

**Which issue(s) this PR fixes**:
Fixes #78

**Special notes for your reviewer**:
tests  in progress.
**Release note**:

```improvement user
CMEK disk encryption is now supported for disks attached to VM. Refer https://github.com/gardener/machine-controller-manager-provider-gcp/blob/master/kubernetes/machine-class.yaml for example
```

### CODE REVIEW

**Comment by ashwani2k** [comment]
[celebrate]     Kumar, Ashwani reacted to your message:
________________________________
From: Tarun Ramakrishna Elankath ***@***.***>
Sent: Thursday, May 25, 2023 6:36:56 AM
To: gardener/machine-controller-manager-provider-gcp ***@***.***>
Cc: Kumar, Ashwani ***@***.***>; Review requested ***@***.***>
Subject: [gardener/machine-controller-manager-provider-gcp] CMEK support changes (PR #84)


What this PR does / why we need it:
Changes required to support CMEK Encrypted disks in GCP.

Which issue(s) this PR fixes:
Fixes #78<https://github.com/gardener/machine-controller-manager-provider-gcp/issues/78>

Special notes for your reviewer:
tests in progress.
Release note:

Support for CMEK Encrypted Disks as described in: https://github.com/gardener/gardener-extension-provider-gcp/issues/564


________________________________
You can view, comment on, or merge this pull request online at:

  https://github.com/gardener/machine-controller-manager-provider-gcp/pull/84

Commit Summary

  *   984b9e7<https://github.com/gardener/machine-controller-manager-provider-gcp/pull/84/commits/984b9e7d142d6ae0e0cd2f659b8ae4725b6d1755> CMEK support changes

File Changes

(5 files<https://github.com/gardener/machine-controller-manager-provider-gcp/pull/84/files>)

  *   M pkg/api/v1alpha1/provider_spec.go<https://github.com/gardener/machine-controller-manager-provider-gcp/pull/84/files#diff-bd1b247473fbbde507d51db370b78fb10ed9413acde9f98c73e290eaf2bfee2a> (9)
  *   M pkg/gcp/machine_controller.go<https://github.com/gardener/machine-controller-manager-provider-gcp/pull/84/files#diff-c529338f56b2f91e80e0fdd2962629f1f69b94d0ac9f7faef21c5791ce7a9b01> (17)
  *   M pkg/gcp/machine_controller_test.go<https://github.com/gardener/machine-controller-manager-provider-gcp/pull/84/files#diff-0219be6431bf8e562f7356c75d8dbeba1f0afcb18f5abfa2243cc4d2b406995f> (150)
  *   M pkg/gcp/machine_controller_util.go<https://github.com/gardener/machine-controller-manager-provider-gcp/pull/84/files#diff-37cdf41d20d791270632d63f8e818d3215cb2ce332bf16c837f3f83d3e1b0f8f> (14)
  *   M pkg/gcp/validation/validation.go<https://github.com/gardener/machine-controller-manager-provider-gcp/pull/84/files#diff-4317df83b48e3723adc9d01622dddff54d3074a2056e258b3ac9a07e41055fb4> (9)

Patch Links:

  *   https://github.com/gardener/machine-controller-manager-provider-gcp/pull/84.patch
  *   https://github.com/gardener/machine-controller-manager-provider-gcp/pull/84.diff

**Comment by elankath** [comment]
# HOW TO MANUALLY TEST FEATURE 
1. Setup a GCP shoot cluster using a service account
1. Setup Cloud KMS Key and Assign IAM role:
    1. Install gcloud CLI: `brew install gcloud`
    1. Login into gcloud: `gcloud auth login`
    1. Create a Cloud KMS keyring: `gcloud kms keyrings create cmektest --location=us-east4`. (This requires IAM privileges to be assigned to you)
    1. Create a Cloud KMS Key for encryption purpose: `gcloud kms keys create alpha --keyring=cmektest --location=us-east4 --purpose encryption`
    1. List to get fully qualified path of Cloud KMS Key: 
    ```
    gcloud kms keys list --keyring=cmektest --location=us-east4
    NAME                                                                                                                                                       PURPOSE
    projects/projectId/locations/us-east4/keyRings/cmektest/cryptoKeys/alpha    ENCRYPT_DECRYPT
    ```
    1. Give the service account used to create the shoot cluster relevant IAM role to use this key
    ```
    gcloud projects add-iam-policy-binding projectId --member serviceAccount:user@projectId.iam.gserviceaccount.com --role roles/cloudkms.cryptoKeyEncrypterDecrypter
    ```
1. Checkout the `machine-controller-manager` and `machine-controller-manager-provider-gcp` in peer directories under `$GOPATH/src/github.com/gardener`.
3. Setup and Start Local MCM: 
     1. `cd $GOPATH/src/github.com/gardener/machine-controller-manager`
     2. `make download-kubeconfigs`
     4. `make start 2>&1 | tee /tmp/mcm.log`
5. Start Local MC 
     1. `cd $GOPATH/src/github.com/gardener/machine-controller-manager-provider-gcp`
     2. `make start 2>&1 | tee /tmp/mcm-gcp.log`
     3. Wait for any pending reconciles to finish.
6. Edit the `MachineClass` yaml
     1. `k edit mcc`
     2.   Modify `providerSpec.disks` as follows: adding `kmsKeyName` and `kmsKeyServiceAccount` as illustrated below (example below makes boot disk CMEK encrypted):
     ```yaml
          - autoDelete: true
             boot: true
             kmsKeyName: projects/projectId/locations/us-east4/keyRings/cmektest/cryptoKeys/alpha
             kmsKeyServiceAccount: user@projectId.iam.gserviceaccount.com
      ```
7.  Trigger restart of `Machine`
      1. List the existing `Running` machine:
       ```
      k get machine
      shoot--userid--shootName-worker-alpha-z1-9485d-xprld
       ```
    1. Add the `force-deletion` label to the machine. (So restart is faster)
        ```
           k label mc shoot--userid--shootName-worker-alpha-z1-9485d-xprld force-deletion=true
        ```
     1. Delete the machine: `k delete mc shoot--userid--shootName-worker-alpha-z1-9485d-xprld`
     1. Wait for reconcile to finish and confirm new machine is `Running`
       ```
       k get machine                                    
       NAME                                                        STATUS    AGE   
       shoot--userid--shootName-worker-alpha-z1-9485d-96fkl   Running   67m   
       ```
 11. Confirm Disk Creation/Encryption.
       1. Test to see that corresponding boot disk is created via gcloud CLI. (The boot disk name is same as machine name)
          ```
             gcloud compute disks list | grep 96fkl
             shoot--userid---shootName-worker-alpha-z1-9485d-96fkl  us-east4-c      zone            50       pd-balanced  READY
           ```
       1. Confirm that disk is using CMEK encryption key:
       ```
           gcloud compute disks describe shoot--shootName-worker-alpha-z1-9485d-96fkl --zone us-east4-c worker-alpha-z1-9485d-96fkl --zone us-east4-c
           creationTimestamp: '2023-05-25T00:46:41.136-07:00'
           diskEncryptionKey:
               kmsKeyName: projects/projectId/locations/us-east4/keyRings/cmektest/cryptoKeys/alpha/cryptoKeyVersions/2
               kmsKeyServiceAccount: user@projectId.iam.gserviceaccount.com
            id: '843731016874110129'
       ```

**Review by kon-angelo — APPROVED** [review]
/lgtm

**Comment by elankath** [comment]
Review comments from meeting:
1. Support empty `kmsKeyServiceAccount`  that defaults to compute engine service agent account. (Google's doc is WRONG here)
2. Move keys to separate `GCPDiskEncryption` struct.

**Comment by ashwani2k** [comment]
A suggestion, shall we move the How to test manually from the https://github.com/gardener/machine-controller-manager-provider-gcp/pull/84#issuecomment-1562553340 to the docs instead as I believe this will be handy later as well.  wdyt?

/hold because of https://github.com/gardener/machine-controller-manager-provider-gcp/pull/84#issuecomment-1562859544

**Comment by elankath** [comment]
Yes, will be moving this inside the `docs` and updating with tests, refactoring.

File: pkg/gcp/machine_controller_util.go [hunk 1]
```diff
@@ -122,6 +123,14 @@ func (ms *MachinePlugin) CreateMachineUtil(ctx context.Context, machineName stri
 				},
 			}
 		}
+		if disk.KmsKeyName != "" {
```
Comment (unmarshall) [inline pkg/gcp/machine_controller_util.go:126]: is there a possibility that someone provides only spaces as the keyName? Safe thing to do would be to just trim `disk.KmsKeyName`

File: pkg/gcp/validation/validation.go [hunk 2]
```diff
@@ -112,6 +113,12 @@ func validateGCPDisks(disks []*api.GCPDisk, fldPath *field.Path) []error {
 		if disk.Boot && "" == disk.Image {
 			allErrs = append(allErrs, field.Required(idxPath.Child("image"), "image is required for boot disk"))
 		}
+		if disk.KmsKeyName != "" && disk.KmsKeyServiceAccount == "" {
```
Comment (unmarshall) [inline pkg/gcp/validation/validation.go:116]: same as above and this time for both `KmsKeyName` and `KmsKeyServiceAccount` as well.

**Review by unmarshall — COMMENTED** [review]
Minor comments

**Review by unmarshall — COMMENTED** [review]
Minor comments

File: pkg/gcp/machine_controller_test.go [hunk 3]
```diff
@@ -70,6 +71,10 @@ const (
...
 	// FailAtNonExistingMachine because existing machine is not found
 	FailAtNonExistingMachine string = "rpc error: code = NotFound desc = Machine with the name \"non-existent-dummy-machine\" not found"
+	// FailAtSpecValidationNoKmsServiceAccount if kmsKeyName specified but kmsKeyServiceAccount missing
+	// FailAtSpecValidationNoKmsServiceAccount if kmsKeyName specified but kmsKeyServiceAccount missing
+	FailAtSpecValidationNoKmsServiceAccount string = "machine codes error: code = [Internal] message = [Create machine \"dummy-machine\" failed on decodeProviderSpecAndSecret: machine codes error: code = [Internal] message = [Error while validating ProviderSpec [spec.disks[0].kmsKeyServiceAccount: Required value: kmsKeyServiceAccount is required to be specified along with kmsKeyName]]]"
+	FailAtSpecValidationNoKmsKeyName        string = "machine codes error: code = [Internal] message = [Create machine \"dummy-machine\" failed on decodeProviderSpecAndSecret: machine codes error: code = [Internal] message = [Error while validating ProviderSpec [spec.disks[0].kmsKeyServiceAccount: Required value: kmsKeyName is required to be specified along with kmsKeyServiceAccount]]]"
```
Comment (ashwani2k) [inline pkg/gcp/machine_controller_test.go:77]: Minor nit pick. 
```suggestion
	// FailAtSpecValidationNoKmsServiceAccount if kmsKeyName is specified but kmsKeyServiceAccount is missing
	FailAtSpecValidationNoKmsServiceAccount string = "machine codes error: code = [Internal] message = [Create machine \"dummy-machine\" failed on decodeProviderSpecAndSecret: machine codes error: code = [Internal] message = [Error while validating ProviderSpec [spec.disks[0].kmsKeyServiceAccount: Required value: kmsKeyServiceAccount is required to be specified along with kmsKeyName]]]"
	// FailAtSpecValidationNoKmsName if kmsKeyServiceAccount is specified but kmsKeyName is missing
	FailAtSpecValidationNoKmsKeyName        string = "machine codes error: code = [Internal] message = [Create machine \"dummy-machine\" failed on decodeProviderSpecAndSecret: machine codes error: code = [Internal] message = [Error while validating ProviderSpec [spec.disks[0].kmsKeyServiceAccount: Required value: kmsKeyName is required to be specified along with kmsKeyServiceAccount]]]"
```

Comment (ashwani2k) [reply pkg/gcp/machine_controller_test.go]: Just for completeness sake that it caught your eye as well, as I saw the commit mentions `Error message grammar`. 
The Lines 74/75 in the original were duplicated without any change to text and were also placed wrongly.

File: pkg/gcp/validation/validation.go [hunk 4]
```diff
@@ -112,6 +114,16 @@ func validateGCPDisks(disks []*api.GCPDisk, fldPath *field.Path) []error {
...
 		}
+		if disk.Encryption != nil {
+			var kmsKeyName = strings.TrimSpace(disk.Encryption.KmsKeyName)
+			var kmsKeyServiceAccount = strings.TrimSpace(disk.Encryption.KmsKeyServiceAccount)
+			if kmsKeyName == "" {
+				allErrs = append(allErrs, field.Required(idxPath.Child("kmsKeyName"), "kmsKeyName is required to be specified"))
+			}
+			if disk.Encryption.KmsKeyServiceAccount != "" && kmsKeyServiceAccount == "" {
+				allErrs = append(allErrs, field.Required(idxPath.Child("kmsKeyServiceAccount"), "kmsKeyServiceAccount should either be explicitly specified or left un-specified to default to the Compute Service Agent"))
+			}
```
Comment (himanshu-kun) [inline pkg/gcp/validation/validation.go:125]: this is the case where customer provides spaces , right?
```suggestion
// to deal with situation where  just spaces have been specified for `kmsKeyServiceAccount`
			if disk.Encryption.KmsKeyServiceAccount != "" && kmsKeyServiceAccount == "" {
				allErrs = append(allErrs, field.Required(idxPath.Child("kmsKeyServiceAccount"), "kmsKeyServiceAccount should either be explicitly specified without spaces or left un-specified to default to the Compute Service Agent"))
			}
```

**Review by himanshu-kun — CHANGES_REQUESTED** [review]
kindly update the file https://github.com/gardener/machine-controller-manager-provider-gcp/blob/15502b8d9a3b50251e250048ccb95cdffdc3c600/kubernetes/machine-class.yaml#L11-L18, and keep the `encryption` section commented (to specify its optional)

Kindly also write `# optional` for `kmsKeyServiceAccount` field  (like done [here](https://github.com/gardener/machine-controller-manager-provider-equinix-metal/blob/a42baed0635e3f73a786d122fca6d7486e255ade/kubernetes/machine-class.yaml#L18))

File: pkg/gcp/machine_controller_util.go [hunk 5]
```diff
@@ -122,6 +123,14 @@ func (ms *MachinePlugin) CreateMachineUtil(ctx context.Context, machineName stri
 				},
 			}
 		}
+		if disk.Encryption != nil {
+			attachedDisk.DiskEncryptionKey = &compute.CustomerEncryptionKey{
+				KmsKeyName:           strings.TrimSpace(disk.Encryption.KmsKeyName),
+				KmsKeyServiceAccount: strings.TrimSpace(disk.Encryption.KmsKeyServiceAccount),
+			}
+			klog.V(3).Infof("(CreateMachineUtil) Set attachedDisk.DiskEncryptionKey.KmsKeyName: %q, attachedDisk.DiskEncryptionKey.KmsKeyServiceAccount: %q",
+				attachedDisk.DiskEncryptionKey.KmsKeyName, attachedDisk.DiskEncryptionKey.KmsKeyServiceAccount)
```
[the remark this answers is not in the document]
Comment (elankath) [reply pkg/gcp/machine_controller_util.go:132]: Ok, it now appears like this in log 
```
I0601 13:09:30.853523   99845 machine_controller_util.go:131] (CreateMachineUtil) For machineName: "shoot--i034796--tre-gcp-alpha-worker-alpha-z1-5bb75-6lvtq", diskLabel: "tre-gcp-alpha", DiskEncryptionKey.KmsKeyName: "projects/sap-se-gcp-k8s-dev-team/locations/us-east4/keyRings/cmektest/cryptoKeys/alpha", DiskEncryptionKey.KmsKeyServiceAccount: "elkat1@sap-se-gcp-k8s-dev-team.iam.gserviceaccount.com"
```

**Review by himanshu-kun — APPROVED** [review]
/lgtm

**Pull request merged**